### PR TITLE
Mesh functors

### PR DESCRIPTION
**Pull request by permcody**
This PR adds the basic ghosting functor capability to MOOSE. I also fixed up some low hanging fruit with "regular_orthogonal_mesh" check that affects a ton of `MeshModifiers` and the phase-field module. There are probably many more but we'll get those as we start doing better DistributedMesh testing.

For testing I added a new GhostUserObject that figures out what elements are geometrically ghosted (NOT algebraically ghosted) so this UO is not useful with ReplicatedMesh as a result. It's just a test object and we can flesh it out more later.

### CODE REVIEW

**Comment by permcody** [comment]
3d_ghost
3d_ghost_2

**Comment by permcody** [comment]
@roystgnr, I just add a few new input file options in MOOSE to turn on the edge and point neighbor functors. I tried for awhile, but couldn't successfully create a test that can verify correct operation of the point neighbor functor. I tried all of the 3D element types and there was no difference for any of the 3D meshes versus just running the problem normally without any functors. Perhaps I'll have to make a mesh manually where there's a node that has several elements attached to it? If you have any good ideas for how I can manufacture a test case, let me know.

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -191,6 +199,21 @@ MooseMesh::MooseMesh(const InputParameters & parameters) :
...
+    if (getParam<bool>("ghost_point_neighbors"))
+      _ghosting_functors.emplace_back(libmesh_make_unique<GhostPointNeighbors>(*_mesh));
+
+    auto num_edge_neighbors = getParam<unsigned short>("num_ghosted_edge_neighbors");
+    if (num_edge_neighbors > 1)
+    {
+      auto default_coupling = libmesh_make_unique<DefaultCoupling>();
+      default_coupling->set_n_levels(num_edge_neighbors);
+      _ghosting_functors.emplace_back(std::move(default_coupling));
+    }
```
Comment (permcody) [inline framework/src/mesh/MooseMesh.C]: @roystgnr - This logic allows me to add both "point" and "edge" functors to the `MeshBase` object. However, when I tested them together, it appears that the second one didn't run (i.e. I received the output below with the thinner banding between partitions). I might be making a mistake somewhere, but I wanted to make sure I was using these correctly before digging in.

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -97,11 +100,16 @@ InputParameters validParams<MooseMesh>()
 
   // Note: This parameter is named to match 'construct_side_list_from_node_list' in SetupMeshAction
   params.addParam<bool>("construct_node_list_from_side_list", true, "Whether or not to generate nodesets from the sidesets (usually a good idea).");
+  params.addParam<unsigned short>("num_ghosted_edge_neighbors", 1, "Parameter to specify the number of geometric edge neighers"
+                                 " that will be available when DistribuedMesh is used. Value is ignored in ReplicatedMesh mode");
+  params.addParam<bool>("ghost_point_neighbors", false, "Boolean to specify whether or not all point neighbors are ghosted"
+                       " when DistribuedMesh is used. Value is ignored in ReplicatedMesh mode");
```
Comment (permcody) [inline framework/src/mesh/MooseMesh.C]: New `[Mesh]` options are here

**Comment by friedmud** [comment]
Noted

**Comment by friedmud** [comment]
Thanks for working on this!

File: framework/src/mesh/MooseMesh.C (hunk 1 above)
Comment (roystgnr) [inline framework/src/mesh/MooseMesh.C]: You'll probably want to be adding these as algebraic coupling functors too, later, but that wouldn't explain reduced ghosting now.  I assume those plots were made by looping over ghosted elements, not over evaluable elements?

**Comment by permcody** [comment]
The logic is really simple and can be seen in the "GhostUserObject" in this
PR. I'm looping over "active elements" and keeping track of any elements
with a different processor ID from the current processor.

**Comment by friedmud** [comment]
@permcody You should easily be able to tell the different between point neighbors and not.  I would think that the test case you have here should show it.  If you're not seeing any difference then maybe something isn't working right...

**Comment by permcody** [comment]
I suppose it's not working then. I ran exodiff on several cases with a
without that new option and exodiff always reported no differences.

**Comment by permcody** [comment]
I know that @jwpeterson has run into this before but Parmetis gives us different partitions on Mac/Linux. It appears that the partitions are consistent on a given platform, just different. For these tests I checked in different gold files for each platform, the input file is the same. Hopefully this passes all tests and can be merged.

File: test/include/userobjects/GhostUserObject.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include "GeneralUserObject.h"
+
+//Forward Declarations
+class GhostUserObject;
+
+template<>
+InputParameters validParams<GhostUserObject>();
+
+/**
+ * An Elemental user object tha uses built-in Random number generation.
```
Comment (friedmud) [inline test/include/userobjects/GhostUserObject.h]: Is it?  Really?

File: framework/src/mesh/MooseMesh.C [hunk 4]
```diff
@@ -96,11 +99,16 @@ InputParameters validParams<MooseMesh>()
 
   // Note: This parameter is named to match 'construct_side_list_from_node_list' in SetupMeshAction
   params.addParam<bool>("construct_node_list_from_side_list", true, "Whether or not to generate nodesets from the sidesets (usually a good idea).");
+  params.addParam<unsigned short>("num_ghosted_edge_neighbors", 1, "Parameter to specify the number of geometric edge neighers"
```
Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]: @permcody and I discussed changing this to `num_ghosted_layers`

**Review by friedmud — DISMISSED** [review]
The testing could use an option to only "view" the ghosts for one particular processor.  Right now you're getting the union of the ghosts for all processors.  Some things could be hiding in there.

**Comment by permcody** [comment]
@jwpeterson - I don't know @friedmud's availability, could you perhaps review this PR? All comments have been addressed. I'd like to get it in along with my other distributed mesh PR (but one of them will need to be rebased after the other since they both made the same change on two lines.

File: framework/src/mesh/MooseMesh.C [hunk 5]
```diff
@@ -96,11 +99,16 @@ InputParameters validParams<MooseMesh>()
 
   // Note: This parameter is named to match 'construct_side_list_from_node_list' in SetupMeshAction
   params.addParam<bool>("construct_node_list_from_side_list", true, "Whether or not to generate nodesets from the sidesets (usually a good idea).");
+  params.addParam<unsigned short>("num_ghosted_layers", 1, "Parameter to specify the number of geometric element layers"
+                                 " that will be available when DistribuedMesh is used. Value is ignored in ReplicatedMesh mode");
+  params.addParam<bool>("ghost_point_neighbors", false, "Boolean to specify whether or not all point neighbors are ghosted"
+                       " when DistribuedMesh is used. Value is ignored in ReplicatedMesh mode");
```
Comment (jwpeterson) [inline framework/src/mesh/MooseMesh.C]: This indentation seems to be off by 1 compared to the indentation on line 111.

File: framework/src/mesh/MooseMesh.C [hunk 6]
```diff
@@ -96,11 +99,16 @@ InputParameters validParams<MooseMesh>()
 
   // Note: This parameter is named to match 'construct_side_list_from_node_list' in SetupMeshAction
   params.addParam<bool>("construct_node_list_from_side_list", true, "Whether or not to generate nodesets from the sidesets (usually a good idea).");
+  params.addParam<unsigned short>("num_ghosted_layers", 1, "Parameter to specify the number of geometric element layers"
+                                 " that will be available when DistribuedMesh is used. Value is ignored in ReplicatedMesh mode");
```
Comment (jwpeterson) [inline framework/src/mesh/MooseMesh.C]: This indentation seems to be off by 1 compared to the indentation on line 111.

File: test/include/auxkernels/GhostAux.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+//Forward Declarations
+class GhostAux;
+class GhostUserObject;
+
+template<>
+InputParameters validParams<GhostAux>();
+
+/**
+ * An AuxKernel that uses built-in Random number generation.
```
Comment (jwpeterson) [inline test/include/auxkernels/GhostAux.h]: Copy-pasted doco is wrong.

File: test/src/auxkernels/GhostAux.C [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+/****************************************************************/
+
+#include "GhostAux.h"
+#include "GhostUserObject.h"
+
+template<>
+InputParameters validParams<GhostAux>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addParam<UserObjectName>("ghost_user_object", "The GhostUserObject where this Aux pulls values from");
```
Comment (jwpeterson) [inline test/src/auxkernels/GhostAux.C]: Please add `classDescription()`

File: test/src/userobjects/GhostUserObject.C [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+#include "MooseMesh.h"
+
+// invalid_processor_id
+#include "libmesh/dof_object.h"
+
+template<>
+InputParameters validParams<GhostUserObject>()
+{
+  InputParameters params = validParams<GeneralUserObject>();
+  params.addParam<unsigned int>("rank", DofObject::invalid_processor_id, "The rank for which the ghosted elements are recorded (Default: ALL)");
```
Comment (jwpeterson) [inline test/src/userobjects/GhostUserObject.C]: Please add `classDescription()`.

File: test/include/userobjects/GhostUserObject.h [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+#include "GeneralUserObject.h"
+
+//Forward Declarations
+class GhostUserObject;
+
+template<>
+InputParameters validParams<GhostUserObject>();
+
+/**
+ * User object to show Ghosted User Objects on ALL processors or a single processor.
```
Comment (jwpeterson) [inline test/include/userobjects/GhostUserObject.h]: Not sure I follow this comment.  Do you mean "show ghosted elements" or "show ghosted geometric entities"?

**Review by jwpeterson — CHANGES_REQUESTED** [review]

**Comment by permcody** [comment]
All comments addressed.

File: test/include/userobjects/GhostUserObject.h [hunk 11]
```diff
@@ -24,7 +24,7 @@ template<>
 InputParameters validParams<GhostUserObject>();
 
 /**
- * User object to show Ghosted User Objects on ALL processors or a single processor.
+ * User object to calculate ghosted elemnts on a single processor or the union across all processors.
```
Comment (jwpeterson) [inline test/include/userobjects/GhostUserObject.h]: "elemnts"

Comment (jwpeterson) [inline test/include/userobjects/GhostUserObject.h]: Other than that, looks good.

**Review by jwpeterson — APPROVED** [review]

**Pull request merged**